### PR TITLE
fabric-admin add ability to grab uid from remote fabric sync device

### PR DESCRIPTION
**Pull request by tehampson**
This allow fabric-admin device to pass MCORE_FS_1_4

With commit https://github.com/project-chip/connectedhomeip/pull/35377/commits/f0b48ef92bf850660ed9f637eba9a6eae8cefde1, confirmed NO regression in:
* ECOINFO_2_1
* ECOINFO_2_2
* BRBINFO_4_1
* CCTRL_2_1
* CCTRL_2_2
* CCTRL_2_3
* MCORE_FS_1_1
* MCORE_FS_1_2
* MCORE_FS_1_3
* MCORE_FS_1_4
* MCORE_FS_1_5

### CODE REVIEW

File: examples/fabric-admin/device_manager/DeviceSynchronization.cpp [hunk 1]
```diff
@@ -128,35 +128,35 @@ void DeviceSynchronizer::OnAttributeData(const ConcreteDataAttributePath & path,
...
         {
-            VerifyOrDie(mController);
-            // TODO(#35333) Figure out how we should recover in this circumstance.
-            CHIP_ERROR err = DeviceSubscriptionManager::Instance().StartSubscription(*mController, mCurrentDeviceData.node_id);
-            if (err != CHIP_NO_ERROR)
+            GetUid(device->GetEndpointId());
+            if (mState == State::GettingUid)
             {
-                ChipLogError(NotSpecified, "Failed start subscription to ");
+                // GetUid was successful and we rely on callback to later on call SynchronizationCompleteAddDevice.
```
Comment (tehampson) [inline examples/fabric-admin/device_manager/DeviceSynchronization.cpp:145]:
```suggestion
                // GetUid was successful and we rely on callback to later call SynchronizationCompleteAddDevice.
```

File: examples/fabric-admin/device_manager/UidGetter.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    {
+        returnValue = CharSpan(mUniqueId);
+    }
+    mOnDoneCallback(returnValue);
+}
+
+void UidGetter::OnDeviceConnected(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    VerifyOrDie(mCurrentlyGettingUid);
+    mClient = std::make_unique<ReadClient>(app::InteractionModelEngine::GetInstance(), &exchangeMgr /* echangeMgr */,
```
Comment (saurabhst) [inline examples/fabric-admin/device_manager/UidGetter.cpp:121]:
```suggestion
    mClient = std::make_unique<ReadClient>(app::InteractionModelEngine::GetInstance(), &exchangeMgr /* exchangeMgr */,
```

**Review by saurabhst — APPROVED** [review]

File: examples/fabric-admin/device_manager/DeviceSynchronization.cpp [hunk 3]
```diff
@@ -205,8 +206,80 @@ void DeviceSynchronizer::StartDeviceSynchronization(chip::Controller::DeviceCont
...
+    VerifyOrDie(DeviceMgr().IsFabricSyncReady());
+    auto remoteBridgeNodeId = DeviceMgr().GetRemoteBridgeNodeId();
+
+    CHIP_ERROR err = mUidGetter.GetUid(
+        [this](std::optional<CharSpan> aUniqueId) {
+            if (aUniqueId.has_value())
+            {
+                this->mCurrentDeviceData.has_unique_id = true;
+                memcpy(this->mCurrentDeviceData.unique_id, aUniqueId.value().data(), aUniqueId.value().size());
+            }
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/DeviceSynchronization.cpp:227]: would the false branch represent an error (we request a UID when one is missing, yet we get nothing)? Should we log an error in this case?

File: examples/fabric-admin/device_manager/DeviceSynchronization.cpp [hunk 4]
```diff
@@ -205,8 +206,80 @@ void DeviceSynchronizer::StartDeviceSynchronization(chip::Controller::DeviceCont
...
+    AddSynchronizedDevice(mCurrentDeviceData);
+    // TODO(#35077) Figure out how we should reflect CADMIN values of ICD.
+    if (!mCurrentDeviceData.is_icd)
+    {
+        VerifyOrDie(mController);
+        // TODO(#35333) Figure out how we should recover in this circumstance.
+        CHIP_ERROR err = DeviceSubscriptionManager::Instance().StartSubscription(*mController, mCurrentDeviceData.node_id);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(NotSpecified, "Failed start subscription to ");
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/DeviceSynchronization.cpp:250]: missing end of log  ` ... to "`

File: examples/fabric-admin/device_manager/DeviceSynchronization.h [hunk 5]
```diff
@@ -65,14 +67,31 @@ class DeviceSynchronizer : public chip::app::ReadClient::Callback
     static DeviceSynchronizer & Instance();
 
 private:
+    enum class State : uint8_t
+    {
+        Idle,             ///< Default state that the object starts out in, where no work has commenced
+        Connecting,       ///< We are waiting for OnDeviceConnected or OnDeviceConnectionFailure callbacks to be called
+        AwaitingResponse, ///< We have started a subscription.
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/DeviceSynchronization.h:74]: would a name of `subscriptionstarted` be better given the description?

We may want to also figure out a name for receivedresponse (like subscription primed?)

File: examples/fabric-admin/device_manager/UidGetter.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    mEndpointId       = endpointId;
+    mOnDoneCallback   = onDoneCallback;
+    mUniqueIdHasValue = false;
+    memset(mUniqueId, 0, sizeof(mUniqueId));
+    mCurrentlyGettingUid = true;
+
+    CHIP_ERROR err = controller.GetConnectedDevice(nodeId, &mOnDeviceConnectedCallback, &mOnDeviceConnectionFailureCallback);
+    if (err != CHIP_NO_ERROR)
+    {
+        mCurrentlyGettingUid = false;
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/UidGetter.cpp:71]: we should probably log the error here as well.

File: examples/fabric-admin/device_manager/UidGetter.cpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+void UidGetter::OnDone(ReadClient * apReadClient)
+{
+    mCurrentlyGettingUid = false;
+    std::optional<CharSpan> returnValue;
+    if (mUniqueIdHasValue)
+    {
+        returnValue = CharSpan(mUniqueId);
+    }
+    mOnDoneCallback(returnValue);
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/UidGetter.cpp:115]: nit:

```suggestion
    mOnDoneCallback(mUniqueIdHasValue ? std::make_optional<CharSpan>(mUniqueId) : std::nullopt);
```

may be more concise (if it compiles).

**Review by yufengwangca — APPROVED** [review]

File: examples/fabric-admin/device_manager/UidGetter.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+    CHIP_ERROR err = mClient->SendRequest(readParams);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to issue subscription to AdministratorCommissioning data");
+        // After calling mOnDoneCallback we are indicating that `this` is deleted and we shouldn't do anything else with
+        // DeviceSubscription.
+        mCurrentlyGettingUid = false;
+        mOnDoneCallback(std::nullopt);
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/UidGetter.cpp:142]: Did you consider calling `OnDone(nullptr)` here? That we we would have a single place changing mCurrentlyGettingUid and mOnDoneCallback, however the downside of having OnDone being passed in a nullptr potentially. We could even do `OnDone(mClient.get())` if we want non-null arguments to it.

File: examples/fabric-admin/device_manager/UidGetter.cpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    }
+}
+
+void UidGetter::OnDeviceConnectionFailure(const ScopedNodeId & peerId, CHIP_ERROR error)
+{
+    VerifyOrDie(mCurrentlyGettingUid);
+    ChipLogError(NotSpecified, "DeviceSubscription failed to connect to " ChipLogFormatX64, ChipLogValueX64(peerId.GetNodeId()));
+
+    mCurrentlyGettingUid = false;
+    mOnDoneCallback(std::nullopt);
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/UidGetter.cpp:153]: same here: maybe OnDone is clearer

File: examples/fabric-admin/device_manager/UidGetter.h [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+#pragma once
+
+#include <app/ReadClient.h>
+#include <controller/CHIPDeviceController.h>
+
+#include <memory>
+#include <optional>
+
+class UidGetter : public chip::app::ReadClient::Callback
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/UidGetter.h:27]: a class doccomment would help in understanding usage.

File: examples/fabric-admin/device_manager/DeviceSynchronization.h (hunk 5 above)
Comment (tehampson) [reply examples/fabric-admin/device_manager/DeviceSynchronization.h:74]: ops comment is copy and paste from somewhere else. I meant to update

File: examples/fabric-admin/device_manager/DeviceSynchronization.cpp [hunk 11]
```diff
@@ -128,35 +128,35 @@ void DeviceSynchronizer::OnAttributeData(const ConcreteDataAttributePath & path,
...
 {
     // Report end is at the end of all attributes (success)
+    MoveToState(State::ReceivedResponse);
+}
+
+void DeviceSynchronizer::OnDone(chip::app::ReadClient * apReadClient)
+{
 #if defined(PW_RPC_ENABLED)
-    if (!DeviceMgr().IsCurrentBridgeDevice(mCurrentDeviceData.node_id))
+    if (mState == State::ReceivedResponse && !DeviceMgr().IsCurrentBridgeDevice(mCurrentDeviceData.node_id))
```
Comment (andy31415) [inline examples/fabric-admin/device_manager/DeviceSynchronization.cpp:137]: would it be possible to somehow reduce the deep nesting in here?
Could be done as a followup - not high priority as logic seems reasonable.

**Review by andy31415 — APPROVED** [review]

File: examples/fabric-admin/device_manager/DeviceSynchronization.cpp (hunk 11 above)
Comment (tehampson) [reply examples/fabric-admin/device_manager/DeviceSynchronization.cpp:137]: I will add in follow up PR https://github.com/project-chip/connectedhomeip/issues/35382

File: examples/fabric-admin/device_manager/UidGetter.h (hunk 10 above)
Comment (tehampson) [reply examples/fabric-admin/device_manager/UidGetter.h:27]: I will add in follow up PR https://github.com/project-chip/connectedhomeip/issues/35382

**Pull request merged**